Tabs to switch devices

xnkevinnguyen opened the pull request:
# Description:
Switch between microbit and cpx device

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Limitations:

Please describe limitations of this PR

# Testing:

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] My code has been formatted with `npm run format` and passes the checks in `npm run check`
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

xnkevinnguyen committed:
Add tabs with fabric library

xnkevinnguyen committed:
Update react version to 16.9.0 to fix issue of typeerror for test renderer

Add basic test for device.tsx

xnkevinnguyen committed:
Tabs are not rerendering components but only hiding/showing pertinent devices component

Linting

Remove unused ressource from device.tsx

Update snapshot for device

sagarmanchanda commented on src/view/container/device/Device.test.tsx:
```diff
@@ -0,0 +1,27 @@
+import { IntlProvider } from "react-intl";
+import Device from "./Device";
+import * as testRenderer from "react-test-renderer";
+
```
You can wrap multiple it blocks within a describe block

sagarmanchanda commented on src/view/container/device/Device.test.tsx:
```diff
@@ -0,0 +1,27 @@
+it("renders correctly", () => {
+    const component = testRenderer
+        .create(
+            <IntlProvider locale="en">
```
Please use consistent line formatting

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
 ];
-
+interface IState {
+    currentDevice: string | undefined;
```
Just having the type string is enough. undefined can be assigned to type string.

xnkevinnguyen replied:
makes sense, I will use ? annotation instead for that property

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+interface IProps {
+    children?: any;
+}
+const DEVICE_LIST_KEY = {
```
enum

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+    constructor(props: IProps) {
+        super(props);
+        this.state = {
+            currentDevice: DEVICE_LIST_KEY.MICROBIT,
```
Did you confirm with PM about this, to have Microbit as the default one?

sagarmanchanda replied:
@sayyedamussa

sayyedamussa replied:
@sagarmanchanda yes we want to make It default

smmatte replied:
We shouldn't make micro:bit the default one until it's full featured.

smmatte replied:
And once micro:bit is full featured, we should not force the change on existing users

xnkevinnguyen replied:
Got it.

sayyedamussa replied:
Sounds good - @sagarmanchanda @xnkevinnguyen  lets not make microbit to default

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+                    onLinkClick={this.handleTabClick}
+                >
+                    <PivotItem
+                        headerText="Micro:bit"
```
Should be abstracted out as a constant somewhere, also might be a localized string?

xnkevinnguyen replied:
definitely, good catch :)

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+                        itemKey={DEVICE_LIST_KEY.MICROBIT}
+                    />
+
+                    <PivotItem headerText="CPX" itemKey={DEVICE_LIST_KEY.CPX} />
```
headerText should be abstracted out

xnkevinnguyen replied:
It is now in constants

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+
+                    <PivotItem headerText="CPX" itemKey={DEVICE_LIST_KEY.CPX} />
+                </Pivot>
+                <div>
```
You could use <React.Fragment> when there's no special need to have a div and you just throw it in for the readability sake

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+                                    : "hidden",
+                        }}
+                    >
+                        microbit
```
I'm not sure what's happening here?

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+                    <div
+                        style={{
+                            visibility:
+                                this.state.currentDevice === DEVICE_LIST_KEY.CPX
```
Doesn't pivot item take care of this?

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
             </div>
         );
     }
+    private handleTabClick = (item: PivotItem | undefined): void => {
```
undefined can be assigned to any type so, you don't have to explicitly call that out

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+    private handleTabClick = (item: PivotItem | undefined): void => {
+        if (item) {
+            this.setState({ currentDevice: item.props.itemKey });
+            console.log(this.state.currentDevice);
```
Do we need this console.log?

xnkevinnguyen replied:
removing it

sagarmanchanda commented on src/view/container/device/Device.tsx:
```diff
@@ -53,17 +54,71 @@ const CPX_TOOLBAR_BUTTONS: Array<{ label: any; image: any }> = [
+        }
+    };
 }
 export default Device;
```
Personally not a big fan of export defaults hah :p

xnkevinnguyen replied:
I will start making them standard export since it will make us avoid renaming

xnkevinnguyen committed:
Fix pr with feedback

Abstract tab in another component

Abstract tab in another component

Abstract tab in another component

Load devices dynamically

Abstract cpx

Update snapshot

xnkevinnguyen committed:
Add tests to components created and modified

Add text in constants

markAtMicrosoft commented on src/view/App.tsx:
```diff
@@ -1,21 +1,39 @@
 // Licensed under the MIT license.
 
-"use strict";
+// "use strict";
```
Unless you plan this to be temporary.  Don't comment, remove. (If temporary, indicate in commented code.
We can always return to previous commits if we need to see what was there.

markAtMicrosoft commented on src/view/constants.ts:
```diff
@@ -6,6 +6,10 @@ export const CONSTANTS = {
     },
+    DEVICE_NAME: {
+        CPX: "CPX",
+        MICROBIT: "Micro:bit",
```
Case, all trademark and marketing info use "micro:bit" no cap...  Do we need to conform?

xnkevinnguyen replied:
Good one, we should definitely conform to their branding

xnkevinnguyen committed:
Update micro:bit displayed word to match branding

xnkevinnguyen committed:
Refactor cpx image to a class component to have proper rerendering

xnkevinnguyen committed:
Update tests snapshots for ui changes

sagarmanchanda commented on src/view/App.spec.tsx:
```diff
@@ -0,0 +1,28 @@
+import * as testRenderer from "react-test-renderer";
+import App from "./App";
+
+describe("App component should", () => {
```
nit: just "App component" here

sagarmanchanda commented on src/view/App.spec.tsx:
```diff
@@ -0,0 +1,28 @@
+import App from "./App";
+
+describe("App component should", () => {
+    it("render correctly", () => {
```
nit: "should render ...."

xnkevinnguyen replied:
noted!

sagarmanchanda commented on src/view/App.tsx:
```diff
@@ -1,21 +1,38 @@
         );
     }
+
+    handleDeviceChange = (item?: PivotItem) => {
```
nit: private

+1 for using arrow functions :D

sagarmanchanda commented on src/view/App.tsx:
```diff
@@ -1,21 +1,38 @@
+}
+
+class App extends React.Component<{}, IState> {
+    state = {
```
I'd put this inside the constructor

this.state = initialState

xnkevinnguyen replied:
Ok, declaring initial state makes it more readable

sagarmanchanda commented on src/view/components/cpx/Svg_utils.tsx:
```diff
@@ -3,6 +3,8 @@
 
 // Adapted from : https://github.com/microsoft/pxt/blob/master/pxtsim/svg.ts
 
+/* tslint:disable */
```
What was the reason to add this?

xnkevinnguyen replied:
That file is currently failing the tslint, but I will fix it

xnkevinnguyen replied:
(unrelated to this pr)

sagarmanchanda commented on src/view/components/tab/Tab.tsx:
```diff
@@ -0,0 +1,27 @@
+interface IProps {
+    handleTabClick: (item?: PivotItem) => void;
+}
+export class Tab extends React.Component<IProps, any> {
```
If it doesn't have a state, consider using React.FC

Also, instead of any, I'd use {} empty state to pass to React.Component

sagarmanchanda reviewed (approved):
Amazing job 🥇 with the initial PR for microbit. Please make the suggested changes before checking it in.

xnkevinnguyen committed:
Merge branch 'dev' into users/t-xunguy/tabs

vandyliu commented on src/view/components/cpx/CpxImage.tsx:
```diff
@@ -20,30 +20,34 @@ interface IProps {
+            setupSwitch(this.props);
+            this.updateImage();
+        } else {
+            console.log("Cannot find svg");
```
Should we do more than just console.log()? Should we even include console.log(). Not very familiar with this situation to be honest

xnkevinnguyen replied:
Removed the console log for now

xnkevinnguyen committed:
Implement feedback from reviews

xnkevinnguyen committed:
Merge branch 'users/t-xunguy/tabs' of https://github.com/microsoft/vscode-python-devicesimulator into users/t-xunguy/tabs

xnkevinnguyen committed:
Remove outdated tests
